[HugeNoob] iP

HugeNoob opened the pull request:
# Introducing LilBro Chatbot 🥶 

> "The key is not to prioritize what's on your schedule, but to schedule your priorities." - Stephen Covey

## Description

In this pull request, we are introducing significant enhancements to our chatbot, LilBro. These enhancements include new features and improvements to its task management capabilities.

## Features Added

### 1. Task Management Commands

- **`todo`**: You can now add simple todo tasks to LilBro. 

- **`deadline`**: For tasks with a specific deadline, use `deadline` followed by the task description and the deadline in the format `by [yyyy-MM-dd HH:mm]`.

- **`event`**: To add tasks with a start and end time, use `event` followed by the task description, `from [yyyy-MM-dd HH:mm]`, and `to [yyyy-MM-dd HH:mm]`.

- **`list`**: You can easily list all your current tasks with the `list` command.

- **`mark`**: Mark tasks as done with the `mark` command, followed by the task ID.

- **`unmark`**: Unmark task as not done using the `unmark` command, followed by the task ID.

- **`delete`**: Remove tasks from your list by using the `delete` command followed by the task ID.

- **`find`**: Search for tasks containing a specific keyword using the `find` command followed by your search query.

### 2. Task Persistence

- **Local Storage**: LilBro now has the ability to store your tasks locally. Your tasks at the end of each session will be saved, allowing you to access them the next time you start LilBro again.

## Usage Examples

Here are some usage examples for the new features:

```bash
# Add a simple todo task
$ todo Buy groceries

# Add a deadline task
$ deadline homework /by 2023-09-05 23:59

# Add an event task
$ event CS2103T Lecture /from 2023-09-08 16:00 /to 2023-09-08 18:00

# List all existing tasks
$ list

# Mark a task as done
$ mark 1

# Unmark a task as done
$ unmark 1

# Remove a specific task
$ delete 1

# Find all tasks containing a query 
$ find CS2103T
```

If you Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Duke {
    public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
    }
}
```

## Requirements

- [x]  Windows/MacOS/Linux
- [x] Java 11

## Quick Start

1. Download Java 11 from [here](https://www.oracle.com/sg/java/technologies/javase/jdk11-archive-downloads.html).
2. Download the latest jar file from the [releases tab](https://github.com/HugeNoob/ip/releases).
3. Run with `java -jar duke.jar` from your terminal.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

HugeNoob committed:
chore: Ignore files generated by JDTLS

HugeNoob committed:
feat: [Level-0] Rename, Greet, Exit

- Rename Duke to LilBro
- Implement initial skeletal version

HugeNoob committed:
feat: [Level-1] Echo

- Echo commands entered by user and exit with "bye" command

HugeNoob committed:
feat: [Level-2] Add, List

- Add ability to store user tasks
- Add ability to display user tasks when requested

HugeNoob committed:
feat: [Level-3] Mark as Done

- Add ability to mark tasks as done/undone

HugeNoob committed:
feat: [Level-4] ToDos, Events, Deadlines

- Add support for tracking three types of tasks

HugeNoob committed:
test: [A-TextUiTesting] Automated Text UI Testing

- Semi-automate testing of LilBro with I/O redirection

HugeNoob committed:
feat: [Level-5] Handle Errors

- Handle various invalid user input cases

HugeNoob committed:
test: [Level-5] Update test cases

- Update test case with new error messages

HugeNoob committed:
feat: [Level-6] Delete

- Add support for deleting tasks
- Fix incorrect check of task indexing

HugeNoob committed:
test: [Level-6] Update test case

- Add tests for delete command

HugeNoob committed:
docs: [Level-6] Add Javadocs for all classes and methods

HugeNoob committed:
feat: [A-Enums] Use Enums

- Add TaskType enum

HugeNoob committed:
feat: Trim user inputs

- Trim task inputs
- Add whitespace test cases to test trimming

HugeNoob committed:
style: Fix switch case indents

HugeNoob committed:
fix: Wrong slash in error messages

HugeNoob committed:
feat: Save

HugeNoob committed:
Merge branch 'branch-Level-7'

HugeNoob committed:
feat: Dates and Times

HugeNoob committed:
test: Update test cases for Level-8

HugeNoob committed:
Merge branch 'branch-Level-8'

HugeNoob committed:
feat: Ui abstraction

HugeNoob committed:
test: Update tests with new Ui formats

HugeNoob committed:
refactor: Rearrange logic of main class

HugeNoob committed:
feat: Add Command classes

HugeNoob committed:
feat: Add Parser class

HugeNoob committed:
Merge branch 'branch-A-MoreOOP'

HugeNoob committed:
refactor: Organize into Packages

HugeNoob committed:
Merge branch 'branch-A-Packages'

HugeNoob committed:
feat: Use Gradle

HugeNoob committed:
feat: Add checkstyle to Gradle

HugeNoob committed:
style: Fix checkstyle errors, except JavaDocs

HugeNoob committed:
Merge branch 'branch-A-Gradle'

HugeNoob committed:
refactor: Improve command constructors

HugeNoob committed:
fix: File rename mistake

HugeNoob committed:
test: Test Parser, Event, and Task

HugeNoob committed:
Merge branch 'branch-A-JUnit'

HugeNoob committed:
feat: Find

HugeNoob committed:
docs: Add JavaDoc comments

HugeNoob committed:
style: Follow the Coding Standard

HugeNoob committed:
Merge branch 'branch-A-CodingStandard'

HugeNoob committed:
Merge branch 'branch-A-JavaDoc'

HugeNoob committed:
Merge branch 'branch-Level-9'

HugeNoob committed:
docs: Add JavaDocs for Find

HugeNoob committed:
chore: Fix checkstyle versioning

HugeNoob committed:
Merge branch 'branch-A-CheckStyle'

HugeNoob committed:
feat: Add basic JavaFX functionality

HugeNoob committed:
refactor: Refactor task list and commands to produce GUI output

HugeNoob committed:
docs: Update JavaDocs for GUI functionality

HugeNoob committed:
Merge branch 'branch-Level-10'

- Shift towards GUI centric output, while retaining CLI output

BREAKING CHANGE: text-ui-test scripts no longer work due to hardcoded
JavaFX paths

rayyan35p commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,78 @@
+    public Duke(String path) {
+        Ui.greet();
+        Ui.println("Checking for a save file...");
+        String projectRoot = System.getProperty("user.dir");
```
Can I ask why you used System.getProperty to get the current working directory?

HugeNoob replied:
I don't recall exactly why I chose to do this, but it seems to be a common method to retrieve the current working directory.

https://stackoverflow.com/questions/4871051/how-to-get-the-current-working-directory-in-java
https://stackoverflow.com/questions/24709769/java-using-system-getpropertyuser-dir-to-get-the-home-directory

rayyan35p commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,78 @@
+    public Duke(String path) {
+        Ui.greet();
+        Ui.println("Checking for a save file...");
+        String projectRoot = System.getProperty("user.dir");
+        String fullPath = Path.of(projectRoot, path).toString();
+        this.storage = new Storage(fullPath);
```
Adding on to the previous comment, perhaps you could possibly do this since you are already using Paths:
```java
Path fullPath= Paths.get(".").resolve(inputPath);
```
`Paths.get(".")` gets the current working directory and you can add on the user input using `resolve(inputPath)`

You can take a look [here](https://docs.oracle.com/javase/tutorial/essential/io/pathOps.html#create) for Path operations :)

HugeNoob replied:
Thanks for the suggestion, I was not aware of `Paths.get(".")`.

rayyan35p commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,100 @@
+     */
+    public Storage(String path) throws DukeException {
+        this.path = path;
+        File file = new File(path);
```
Based on your code in Duke.java, I thought you were planning on using the java.nio package(which from what I read, is newer and better than java.io which is what this line is using. 

I actually tried using Paths entirely initially but I couldn't find many tutorials online for nio classes so I stuck to using only java.io.File classes. 

Since you already incorporated some nio classes like Path and Files into your code maybe you can consider using those mainly instead?

You can try taking a look at [this](https://www.baeldung.com/java-how-to-create-a-file)

rayyan35p commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,176 @@
+public class TaskList {
+
+    /** The list that is used to store the user's tasks. */
+    private ArrayList<Task> list;
```
Would it be better to name it "tasks"? Instead of list

rayyan35p reviewed (commented):
LGTM! Good code quality throughout the entire code, I only managed to comment about some stuff unrelated to the coding standard because there wasn't many issues 👍

nubnubyas commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,79 @@
+     * @return The command that was parsed.
+     * @throws DukeInvalidCommandException If the given input is invalid.
+     */
+    public Command parse(String input) throws DukeInvalidCommandException {
+        String[] parts = input.split(" ", 2);
+        CommandType commandType = CommandType.fromString(parts[0]);
+        String commandArgs = parts.length > 1 ? parts[1].trim() : "";
+
+        switch (commandType) {
+        case EXIT:
+            return new ExitCommand(this.taskList, commandArgs);
+
+        case LIST_TASKS:
+            return new ListTasksCommand(this.taskList, commandArgs);
+
+        case ADD_TODO:
+            return new AddToDoCommand(this.taskList, commandArgs);
+
+        case ADD_DEADLINE:
+            return new AddDeadlineCommand(this.taskList, commandArgs);
+
+        case ADD_EVENT:
+            return new AddEventCommand(this.taskList, commandArgs);
+
+        case MARK_TASK:
+            return new MarkTaskCommand(this.taskList, commandArgs);
+
+        case UNMARK_TASK:
+            return new UnmarkTaskCommand(this.taskList, commandArgs);
+
+        case DELETE_TASK:
+            return new DeleteTaskCommand(this.taskList, commandArgs);
+
+        case FIND_TASK:
+            return new FindTaskCommand(this.taskList, commandArgs);
+
+        default:
+            throw new DukeInvalidCommandException("I'm gonna be honest, no idea what you're saying.");
+        }
```
This looks really clean. Good job for doing the stretch goals.

nubnubyas commented on src/test/java/duke/parser/ParserTest.java:
```diff
@@ -0,0 +1,41 @@
+package duke.parser;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import org.junit.jupiter.api.Test;
+
+import duke.commands.Command;
+import duke.commands.CommandType;
+import duke.exceptions.DukeInvalidCommandException;
+import duke.tasks.TaskList;
```
Good spacing and ordering of imported packages. I like how you also separated static imports from normal ones. Good job for doing the optional Checkstyle for IP (I assume you did :>).

nubnubyas commented on src/main/java/duke/ui/MainWindow.java:
```diff
@@ -0,0 +1,104 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import duke.Duke;
+import duke.commands.Command;
+import duke.commands.CommandType;
+import duke.exceptions.DukeException;
+import javafx.fxml.FXML;
+import javafx.scene.control.Button;
+import javafx.scene.control.ScrollPane;
+import javafx.scene.control.TextField;
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.Region;
+import javafx.scene.layout.VBox;
```
Perhaps you can separate the imports from duke and the ones from javafx? Would improve readability.

nubnubyas reviewed (commented):
LGTM
Your code quality is really good. I only managed to find a small improvement throughout. Good job on doing optional tasks too. 👍

HugeNoob committed:
ci: Add GitHub CI to the project

Gradle checks and tests are run manually, which is not ideal.
Forgetting to run checks for minor changes will introduce buggy code.

This CI script runs ./gradlew check on each push and PR, which will
help maintain code quality.

HugeNoob committed:
Merge pull request #1 from HugeNoob/branch-A-CI

ci: Add GitHub CI to the project

HugeNoob committed:
feat: Add assertions

* Assert task descriptions to be non-empty
* Assert that save file exists before storing tasks

HugeNoob committed:
refactor: Improve code quality

HugeNoob committed:
feat: Use streams

HugeNoob committed:
Merge pull request #2 from HugeNoob/branch-A-Assertions

feat: Add assertions

HugeNoob committed:
Merge branch 'master' into branch-A-CodeQuality

HugeNoob committed:
Merge pull request #3 from HugeNoob/branch-A-CodeQuality

refactor: Improve code quality

HugeNoob committed:
Merge branch 'master' into branch-A-Streams

HugeNoob committed:
Merge pull request #4 from HugeNoob/branch-A-Streams

feat: Use streams

HugeNoob committed:
feat: Friendlier syntax for commands

User has to type out full command words.

I'm a lazy Neovim user and hate typing verbose commands.

Add some shorter aliases for all commands.

New mappings (while retaining old ones):
* "exit": EXIT
* "q": EXIT
* "ls": LIST_TASKS
* "td": ADD_TODO
* "dl": ADD_DEADLINE
* "ev": ADD_EVENT
* "m": MARK_TASK
* "u": UNMARK_TASK
* "rm": DELETE_TASK
* "f": FIND_TASK

HugeNoob committed:
Merge pull request #5 from HugeNoob/branch-C-FriendlierSyntax

feat: Friendlier syntax for commands

HugeNoob committed:
fix: Rename image files to fix jar build

HugeNoob committed:
docs: Fix header comments identified by se-bot

HugeNoob committed:
docs: Add finalised Ui image

HugeNoob committed:
fix: Minor string formatting issues

HugeNoob committed:
docs: Add User Guide
